# SoftMax Activation Function

### Initializing Data
- We initialise the scores directly supposing that we obtain them with F(x)= W.x,where W is the matrix of weights and x is the matrix of inputs

In [1]:
import numpy as np
cat_image_scores=np.asarray([3.2,5.1,-1.7])
car_image_scores=np.asarray([1.3,4.9,2.0])
frog_image_scores=np.asarray([2.2,2.5,-3.1])

### Structuring Data with Pandas
- Visialisation if the data with pandas dataframe

In [2]:
import pandas as pd
df=pd.DataFrame([cat_image_scores,car_image_scores,frog_image_scores],columns=['cat_class','carclass','frog_class'],index=['Cat_image','Car_image','Frog_image'])
scores_array=np.asarray(df)
print(scores_array)
df.head()

[[ 3.2  5.1 -1.7]
 [ 1.3  4.9  2. ]
 [ 2.2  2.5 -3.1]]


cat_class  carclass  frog_class
Cat_image         3.2       5.1        -1.7
Car_image         1.3       4.9         2.0
Frog_image        2.2       2.5        -3.1

### SoftMax Activation Function
- The softmax activation function which is given by <br>
<b>softmax(z)_i=(exp(z_i))/(sum(exp(z_j)))</b>


In [3]:
from math import exp
def SoftMax(sca):
    i=0
    denomin=0
    nomin=0
    probs=[]
    if sca.ndim <= 1:
        denomin=sum(np.exp(sca))
        for k in sca:
            nomin=exp(k)
            probs.append(nomin/denomin)
    else:
        while i<len(sca):
            prob=[]
            denomin=sum(np.exp(sca[i])) 
            for j in sca[i]:
                nomin=exp(j)
                prob.append(nomin/denomin)
            probs.append(prob)
            i+=1
    return probs

### Application of Softmax
- The Softmax activation function recieves the final scores of the neural networks(on last layer in general) and converts them into probabilities.
- The bigger the score is the bigger the probabilities of the input to belong to a certain class of outputs.
- Each element of the array below contain a set of probablities of a certain image to belong to ecery class of images.<br>
    for example the first element :[0.12998253901717313, 0.8690495352082657, 0.0009679257745610439]
- The image belong the class with the biggest probabilitie.<br>
    in the above element the image belong to the seconde class.

In [10]:
sm=SoftMax(scores_array)
print('this are the probablities of each image')
sm

this are the probablities of each image


[[0.12998253901717313, 0.8690495352082657, 0.0009679257745610439],
 [0.025244883478939707, 0.9239181640655688, 0.05083695245549154],
 [0.42465542582674165, 0.5732248667371652, 0.0021197074360931266]]

## Cross-Caterogical-Entropy
- this the loss function usually used for mesuring the multiclass lassification problemes.
- It has a simple principe, it mesures how strong is the probabilitie of the image to be in the right class.<br>
    * in the previous example, suppose that the postive class is the first class(cat_image_class): <br>
    [0.12998253901717313, 0.8690495352082657, 0.0009679257745610439]<br>
    * The SoftMax_loss function "-log(P)" would mesure the strenght of the first probabilitie (0.12998...),where it converts the probabilitie to a mesurable value.

In [5]:
from math import log
def SoftMax_loss(sca):
    ss=np.asarray(SoftMax(sca))
    loss=[]
    if ss.ndim <= 1 :
        loss= -np.log(ss[0])
    else:
        diag=np.diagonal(ss)
        for i in diag:
            loss.append(-np.log10(i))
    return loss

### The losses of each input

In [12]:
print('The loss of each positive class:')
sml=SoftMax_loss(scores_array)
print(sml)
print('the loss of all the positive classes:')
fsml=sum(sml)/len(sml)
print(fsml)

The loss of each positive class:
[0.8861149839839801, 0.03436649465512438, 2.6737240766459403]
the loss of all the positive classes:
1.1980685184283482
